duckduckgo/iOS
# Autofill "Never Save for this Site" feature

Task/Issue URL: https://app.asana.com/0/0/1205783642285427/f
Tech Design URL: https://app.asana.com/0/481882893211075/1205590957854802/f
CC:

**Description**:
Adds the functionality to allow users to choose to "Never Save for this Site" from the autofill save login prompt, with the option to reset this list from the Logins screen.



**Steps to test this PR**:
Note - I'm still waiting on translations so that will probably be a separate PR

1. Visit https://fill.dev/form/login-simple and login using new credentials. In the `Save Login` prompt, confirm the bottom button is now `Never Ask for This Site` 
<img width='200' src="https://github.com/duckduckgo/iOS/assets/91189922/8d991d91-ddda-423c-ac45-3882a3b2b3fc"/>

2. Tap it and confirm this pixel is fired:` m_autofill_logins_save_login_exclude_site_confirmed` 
3. Go back to https://fill.dev/form/login-simple
4. Login again with new credentials and confirm that you no longer get the `Save Login` prompt
5. Visit another site and, logging with new credentials, confirm the `Save Login` prompt is still presented for other sites
6. Go to https://fill.dev/form/registration-email and confirm that `Password Generation` is disabled for this site (no key in the password field and tapping in the field does not trigger the generated password prompt) -- this is why we needed to adopt support for `getRuntimeConfiguration` now
7. Visit another website with a registration form and confirm that `Password Generation` is still enabled for it 
8. Go the Logins screen and you should see a new row with the option to `Reset Excluded Sites` <img width='400' src="https://github.com/duckduckgo/iOS/assets/91189922/43bc5dd6-106a-46cc-994b-6434327645ce"/>

9. Tap the row and confirm this pixel is fired: `m_autofill_settings_reset_excluded_displayed`
10. Tap `Cancel` in the action sheet that was presented and confirm this pixel is fired `m_autofill_settings_reset_excluded_dismissed`
11. Tap the row and choose `Reset Excluded Sites` and confirm this pixel is fired: `m_autofill_settings_reset_excluded_confirmed`
12. The `Reset Excluded Sites` row should now be gone (it is only added to the screen when a user has at least 1 site saved as `Never Prompt`)
13. Go back to https://fill.dev/form/login-simple and confirm the Save Login prompt now appears again 
14. Final test case (a small bug fix): confirm Private Duck Addresses are not autosaved when Autofill is disabled 🙂 



**Copy Testing**:


**Orientation Testing**:


**Device Testing**:


**OS Testing**:


**Theme Testing**:


---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
